massa-pool-worker/src/denunciation_pool.rs

```diff
@@ -1,51 +1,171 @@
-use massa_models::denunciation::{Denunciation, DenunciationIndex};
-use massa_models::prehash::PreHashMap;
+use std::collections::{hash_map::Entry, HashMap};
+use tracing::{debug, info, warn};
 
+use massa_models::denunciation::DenunciationIndex;
+use massa_models::{
+    address::Address,
+    denunciation::{Denunciation, DenunciationPrecursor},
+    timeslots::get_closest_slot_to_timestamp,
+};
 use massa_pool_exports::PoolConfig;
+use massa_pos_exports::SelectorController;
+use massa_time::MassaTime;
 
 pub struct DenunciationPool {
     /// configuration
     config: PoolConfig,
-    /// storage
-    // storage: Storage,
+    /// selector controller to get draws
+    pub selector: Box<dyn SelectorController>,
     /// last consensus final periods, per thread
     last_cs_final_periods: Vec<u64>,
-    /// internal cache
-    denunciations_cache: PreHashMap<DenunciationIndex, Denunciation>,
+    /// Internal cache for denunciations
+    denunciations_cache: HashMap<DenunciationIndex, DenunciationStatus>,
 }
 
 impl DenunciationPool {
-    pub fn init(config: PoolConfig) -> Self {
+    pub fn init(config: PoolConfig, selector: Box<dyn SelectorController>) -> Self {
         Self {
             config,
+            selector,
             last_cs_final_periods: vec![0u64; config.thread_count as usize],
             denunciations_cache: Default::default(),
         }
     }
 
     /// Get the number of stored elements
     pub fn len(&self) -> usize {
-        self.denunciations_cache.len()
+        self.denunciations_cache
+            .iter()
+            .filter(|(_, de_st)| matches!(*de_st, DenunciationStatus::DenunciationEmitted(..)))
+            .count()
     }
 
-    // Not used yet
-    /*
     /// Checks whether an element is stored in the pool
-    pub fn contains(&self, id: &DenunciationId) -> bool {
-        self.storage.get_denunciation_refs().contains(id)
+    pub fn contains(&self, denunciation: &Denunciation) -> bool {
```

Comment:
Equality comparison on whole structures is usually not a good practice. Better performance can be achieved by just taking the DenunciationIndex as argument, no need to iterate

Reply:
It is used in unit tests so I've put this behind a test flag.